fms/modules/embedding.py

```diff
@@ -1,21 +1,129 @@
-import math
-from typing import List, Tuple, Union
+from typing import Optional
 
 import torch
 import torch.nn as nn
-from numpy import sign
 from torch.distributed.distributed_c10d import ProcessGroup
 
 from fms.distributed.tensorparallel import (
     all_gather_from_tensor_model_parallel_region,
     apply_colwise_tp,
     apply_embedding_tp,
-    apply_rowwise_tp,
     copy_to_tensor_model_parallel_region,
-    reduce_from_tensor_model_parallel_region,
 )
 
 
+class AbsolutePositionEmbedding(nn.Module):
+    """Special form of embedding that includes a position encoding"""
+
+    def __init__(
+        self,
+        vocab_size: int,
+        emb_dim: int,
+        padding_idx: Optional[int] = None,
+        max_pos: int = 512,
+    ):
+        """
+        Initialize an AbsolutePositionEmbedding
+
+        Parameters
+        ----------
+        vocab_size: int
+            the length of the vocabulary
+        emb_dim: int
+            dimensionality of latent space
+        padding_idx: int, optional
+            Padding token index in the vocabulary. Typically, the token for
+            which positions will be zeroed out. If negative or None, no positions
+            will be zeroed. (default is None)
+        max_pos: int
+            the maximum possible sequence length (default is 512)
+        """
+        super().__init__()
+        self.max_pos = max_pos
+        self.emb_dim = emb_dim
+
+        # if negative padding_idx is given, just set to None
```

Comment:
why do we do this?
this seems like it could cause bugs, and deviates from expected pytorch behavior.
nn.Embedding will raise an error if out of range (or use negative indexing if negative and within range).

Reply:
Yes, I think this can be removed. Just keep it as is, if it is None, we don't do any checking, if it is negative, just allow it through and fall down to pytorch.